components/mediation-monitor/mediation-data-publisher/org.wso2.carbon.das.messageflow.data.publisher/src/main/java/org/wso2/carbon/das/messageflow/data/publisher/publish/DataBridgePublisher.java

```diff
@@ -0,0 +1,76 @@
+package org.wso2.carbon.das.messageflow.data.publisher.publish;
+
+import org.apache.log4j.Logger;
+import org.wso2.carbon.base.ServerConfiguration;
+import org.wso2.carbon.das.data.publisher.util.DASDataPublisherConstants;
+import org.wso2.carbon.das.messageflow.data.publisher.services.MediationConfigReporterThread;
+import org.wso2.carbon.databridge.agent.AgentHolder;
+import org.wso2.carbon.databridge.agent.DataPublisher;
+import org.wso2.carbon.databridge.agent.exception.DataEndpointAgentConfigurationException;
+import org.wso2.carbon.databridge.agent.exception.DataEndpointAuthenticationException;
+import org.wso2.carbon.databridge.agent.exception.DataEndpointConfigurationException;
+import org.wso2.carbon.databridge.agent.exception.DataEndpointException;
+import org.wso2.carbon.databridge.commons.exception.TransportException;
+import org.wso2.carbon.databridge.commons.Event;
+import org.wso2.carbon.utils.CarbonUtils;
+
+public class DataBridgePublisher {
+
+    private static DataPublisher publisher;
+    private static Logger log = Logger.getLogger(DataBridgePublisher.class);
+    private static String receiverUrl;
+    private static String authUrl;
+    private static String username;
+    private static String password;
+
+
+    private synchronized static DataPublisher initDataPublisher(){
+        if (publisher == null){
+            try {
+                loadConfigs();
+
+                publisher = new DataPublisher(null,receiverUrl, authUrl, username, password);
```

Comment:
If the publisher creation fails due to incorrect configurations, each publish call will fail with a log error. IMO, it would be better if we could throw the exceptions and interrupt the publisher creation so that the error will not continue. This could be avoided if we maintain instances of DataBridgePublisher in ConfigurationPublisher and StatisticsPublisher and initialize the instance during the respective holders' initialization.

Reply:
Configuration errors are irrecoverable and have to be restarted. Therefore, let's throw exceptions for configuration exceptions. And for other exceptions, we can log and let Databridge recover such as connection exceptions.

Reply:
In a second thought, this error will get printed every time when the publisher is initialized if there's an error on the config. So, I guess it's okay to let the error get printed since it's critical has to be restarted to fix. Even if we throw an exception there's nothing that can be done to recover from the error by the users of the data publisher instance.

Reply:
On second thought. if there's a configuration error it will get printed every time when it tries to initialize it. Since this is a critical error which requires a restart I guess it's okay to print the error every time when publishing is attempted. Even if we throw an error, no one can catch the error and take any action rather than silently drop the message to be published. Having, multiple data bridge instances are redundant and it will introduce other complications on performance and configs, etc..

Reply:
That's the probem, it'll keep logging errors and and there is a chance that users might ignore it since the main funcationality is not affected. But if we can create the publisher in while initialization, we can interrupt the server startup itself. This is a safe approach don't you think?

Reply:
If the user doesn't want to use analytics I think he has to disable it. Let the server be silent with a wrong config is not correct IMO. I also think it's best if we can do this at the initialization, but since this is a singleton static object which can be used by anyone. Therefore, it's a bit hard to identify an initialization point. That's the reason for initializing the publisher in the first getPublisherCall.

Reply:
That's why we need to remove the singleton object. Anyway, since there is a concern on the connection count, let's leave it as it is.

Reply:
Maintaining a minimal number of connections is important as we have seen many issues with too many connections. A single data bridge instance can have a large footprint with its internal buffers, outbound connection pools, etc. Besides IMO it's not wrong to keep complaining about an irrecoverable problem.